Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -58,17 +60,19 @@ struct HTTPRequestStateMachine {
     }
 
     fileprivate enum ResponseState {
-        /// A sub state for receiving a response. Stores whether the consumer has either signaled demand for more data or
-        /// is busy consuming the so far forwarded bytes
+        /// A sub state for receiving a response. Stores whether the consumer has either signaled demand and whether the
+        /// channel has issued `read` events.
         enum ConsumerControlState {
-            /// the state machine is in this state once it has passed down a request head or body part. If a read event
-            /// occurs while in this state, the readPending flag will be set true. If the consumer signals more demand
-            /// by invoking `forwardMoreBodyParts`, the state machine will forward the read event.
-            case downstreamIsConsuming(readPending: Bool)
-            /// the state machine is in this state once the consumer has signaled more demand by invoking
-            /// `forwardMoreBodyParts`. If a read event occurs in this state the read event will be forwarded
-            /// immediately.
-            case downstreamHasDemand
+            /// The state machines expects further writes to `channelRead`. The writes are appended to the buffer.
+            case waitingForBytes(CircularBuffer<ByteBuffer>)
+            /// The state machines expects a call to `demandMoreResponseBodyParts` or `read`. The buffer is
+            /// empty. It is preserved for performance reasons.
+            case waitingForReadOrDemand(CircularBuffer<ByteBuffer>)
+            /// The state machines expects a call to `read`. The buffer is empty. It is preserved for performance reasons.
+            case waitingForRead(CircularBuffer<ByteBuffer>)
+            /// The state machines expects a call to `demandMoreResponseBodyParts`. The buffer is empty. It is
+            /// preserved for performance reasons.
+            case waitingForDemand(CircularBuffer<ByteBuffer>)
```

Comment:
By making this a mini-state-machine, I recommend going the whole way: pull this out to a file and give it a bunch of methods that correspond to the events that can occur to it (e.g. `func read`). Then, the bigger state machine can just call those.